src/traits.rs

```diff
@@ -134,18 +135,152 @@ impl<T, Rhs, Output> ScalarMul<Rhs, Output> for T where T: Mul<Rhs, Output = Out
+        }
+
+        let hash = poseidon_hash(&mut cs.namespace(|| "hash"), z_vec.clone(), p)?;
+
+        cs.enforce(
+          || "hash = z",
+          |lc| lc + z.get_variable(),
+          |lc| lc + CS::one(),
+          |lc| lc + hash.get_variable(),
+        );
+
+        let inner_output = self.synthesize_step_inner(&mut cs.namespace(|| "inner"), z_vec)?;
+
+        let output_hash = poseidon_hash(&mut cs.namespace(|| "output"), inner_output, p)?;
+
+        Ok(output_hash)
+      }
+      IO::Empty(p) => {
+        let arity = A::to_usize();
+        let mut z_vec = Vec::with_capacity(arity);
+
+        for i in 0..arity {
+          let allocated = AllocatedNum::alloc(cs.namespace(|| format!("z[{}]", i)), || {
+            Err(SynthesisError::AssignmentMissing)
+          })?;
+          z_vec.push(allocated);
+        }
+
+        let hash = poseidon_hash(&mut cs.namespace(|| "hash"), z_vec.clone(), p)?;
+
+        cs.enforce(
+          || "hash = z",
+          |lc| lc + z.get_variable(),
+          |lc| lc + CS::one(),
+          |lc| lc + hash.get_variable(),
+        );
+
+        let inner_output = self.synthesize_step_inner(&mut cs.namespace(|| "inner"), z_vec)?;
+
+        let output_hash = poseidon_hash(&mut cs.namespace(|| "output"), inner_output, p)?;
+
+        Ok(output_hash)
+      }
+    }
+  }
+
+  /// Synthesize the circuit for a computation step with multiple inputs and outputs.
+  /// This method must be implemented if F is a non-unary function.
+  fn synthesize_step_inner<CS: ConstraintSystem<F>>(
+    &self,
+    _cs: &mut CS,
+    _z_vec: Vec<AllocatedNum<F>>,
+  ) -> Result<Vec<AllocatedNum<F>>, SynthesisError> {
+    unimplemented!()
+  }
+
+  /// Return the IO value
+  fn io<'a>(&'a self) -> &'a IO<F, A> {
+    unimplemented!()
+  }
+}
+
+/// Enum for holding the computation F's input/output value(s).
+#[derive(Clone)]
+pub enum IO<'a, F: PrimeField, A: Arity<F>> {
+  /// Unary input/output value
+  Val(F),
+  /// Non-unary input/output values
+  Vals(Vec<F>, &'a PoseidonConstants<F, A>),
+  /// Placeholder for constructing R1CS shape
+  Empty(&'a PoseidonConstants<F, A>),
+}
 
-  /// Execute the circuit for a computation step and return output
-  fn compute(&self, z: &F) -> F;
+impl<'a, F: PrimeField, A: Arity<F>> IO<'a, F, A> {
+  /// Return a single F corresponding to this `IO`.
+  pub fn output(&self) -> F {
+    match self {
+      Self::Val(val) => *val,
+      Self::Vals(vals, p) => {
+        let mut hasher = Poseidon::<F, A>::new_with_preimage(vals, p);
+        hasher.hash()
+      }
+      Self::Empty(_) => unreachable!(),
+    }
+  }
+
+  /// Return a new `IO::Vals`, sharing this `IO`'s poseidon constants
+  pub fn new_vals(&self, vals: Vec<F>) -> Self {
+    match self {
+      IO::Val(_) => unreachable!(),
+      IO::Vals(_, p) => IO::Vals(vals, p),
+      IO::Empty(_) => unreachable!(),
+    }
+  }
+}
+
+/// A helper trait for computing a step of the incremental computation (i.e., F itself)
```

Comment:
Yes, a caller might want more control, in which case they can implement Iterator directly. This will bypass the iterator automatically created using `ComputeStep`. That is also why `prove_with_iterators` is public. When the caller implements an iterator, they can bypass `prove` and call the lower-level function.